File: bindings/zig/src/runtime.zig
//! Profile records (inspect / register / lookup / profiles), Go
//! runtime knobs, and registry diagnostics.
//!
//! A profile record is the JSON object libitb3 accepts in `register`,
//! returns from `lookup` / `inspect`, and embeds in every blob.
//! Optional keys are omitted when empty / zero. The binding treats
//! the record as an opaque string; every field rule is enforced by
//! libitb3.

const std = @import("std");
const ffi = @import("ffi.zig").c;
const err = @import("error.zig");

const Allocator = std.mem.Allocator;

/// Decodes the profile record embedded in `blob_bytes` without
/// constructing a Pipeline. Caller owns the result — release with
/// `allocator.free`.
pub fn inspect(allocator: Allocator, blob_bytes: []const u8) err.Error![]u8 {
    var out: [*c]u8 = null;
    const rc = ffi.itb_inspect(blob_bytes.ptr, blob_bytes.len, &out);
    defer ffi.itb_string_free(out);
    try err.check(rc);
    return allocator.dupe(u8, std.mem.span(out));
}

/// Registers a user-defined Triple profile under `name` from a
/// profile JSON record (a non-empty "name" key inside the record must
/// equal `name`). A duplicate name fails with `error.ProfileExists`.
pub fn register(name: [:0]const u8, profile_json: [:0]const u8) err.Error!void {
    try err.check(ffi.itb_register(name.ptr, profile_json.ptr));
}

/// The profile registered under `name` as its JSON record. An
/// unregistered name fails with `error.UnknownProfile`. Caller owns
/// the result.
pub fn lookup(allocator: Allocator, name: [:0]const u8) err.Error![]u8 {
    var out: [*c]u8 = null;
    const rc = ffi.itb_lookup(name.ptr, &out);
    defer ffi.itb_string_free(out);
    try err.check(rc);
    return allocator.dupe(u8, std.mem.span(out));
}

/// The sorted list of every registered profile name as a JSON array
/// of strings. Caller owns the result.
pub fn profiles(allocator: Allocator) err.Error![]u8 {
    var out: [*c]u8 = null;
    const rc = ffi.itb_profiles(&out);
    defer ffi.itb_string_free(out);
    try err.check(rc);
    return allocator.dupe(u8, std.mem.span(out));
}

/// The shipped hash-primitive registry in canonical order as a JSON
/// array of strings. The list is what a profile's inner-hash name is
/// resolved against, so a caller validates a primitive name without
/// constructing a Pipeline. Caller owns the result.
pub fn hashNames(allocator: Allocator) err.Error![]u8 {
    var out: [*c]u8 = null;
    const rc = ffi.itb_hash_names(&out);
    defer ffi.itb_string_free(out);
    try err.check(rc);
    return allocator.dupe(u8, std.mem.span(out));
}

/// The libitb3 library version string ("" if libitb3 misbehaves).
/// Thread-local C-side buffer — copy before the next call if kept.
pub fn version() [:0]const u8 {
    const v = ffi.itb_version();
    return if (v != null) std.mem.span(v) else "";
}

/// Sets the Go runtime's soft heap limit in bytes; returns the
/// previous limit. A negative value queries without changing.
pub fn setMemoryLimit(bytes: i64) i64 {
    return ffi.itb_set_memory_limit(bytes);
}

/// Sets the Go GC trigger percentage; returns the previous value. A
/// negative value queries without changing.
pub fn setGcPercent(pct: i32) i32 {
    return ffi.itb_set_gc_percent(pct);
}

/// Sets the Go runtime's GOMAXPROCS; returns the previous value. Zero
/// or a negative value queries without changing.
pub fn setGomaxprocs(n: i32) i32 {
    return ffi.itb_set_gomaxprocs(n);
}

/// Writes the Go runtime's heap profile (pprof format) to `path` after
/// one forced garbage collection. An empty path falls back to the
/// ITB_MEMPROFILE environment variable; a path that is still empty, or
/// a file-system failure, is `error.BadInput` with the diagnostic in
/// `lastError`.
pub fn writeHeapProfile(path: [:0]const u8) err.Error!void {
    try err.check(ffi.itb_write_heap_profile(path.ptr));
}

/// The number of i64 slots `poolStats` fills. Size a buffer from this
/// call, never from a constant.
pub fn poolStatsLen() usize {
    return ffi.itb_pool_stats_len();
}

/// The library's pool hit / miss counters. Every counter is a
/// monotonically increasing total since library load — difference two
/// snapshots. Slot layout, with T the tier count in slot 0: tier i
/// holds starter width, checkouts, constructor misses, regrow
/// replacements and bytes allocated at slots 1 + 5*i .. 1 + 5*i + 4;
/// the scratch byte pool's get / new / regrow / regrow-bytes follow at
/// 1 + 5*T, and the parallax chunk pool's at 1 + 5*T + 4. Caller owns
/// the result.
pub fn poolStats(allocator: Allocator) err.Error![]i64 {
    const cap = poolStatsLen();
    const slots = try allocator.alloc(i64, cap);
    errdefer allocator.free(slots);
    if (cap == 0) return slots;
    var written: usize = 0;
    try err.check(ffi.itb_pool_stats(slots.ptr, cap, &written));
    return allocator.realloc(slots, written) catch slots[0..written];
}

File: bindings/rust/src/runtime.rs
//! Process-wide Go runtime knobs, runtime diagnostics, and the library
//! version string.

use std::ffi::{CString, c_char, c_int};
use std::path::Path;

use crate::error::{ItbError, ItbResult, check};
use crate::ffi;
use crate::status::ItbStatus;

/// Sets the Go runtime's soft heap limit in bytes and returns the
/// previous limit. A negative value queries without changing.
pub fn set_memory_limit(bytes: i64) -> ItbResult<i64> {
    let s = ffi::syms()?;
    // SAFETY: plain value-in / value-out call.
    Ok(unsafe { (s.ITB_SetMemoryLimit)(bytes) })
}

/// Sets the Go GC trigger percentage and returns the previous value.
/// A negative value queries without changing.
pub fn set_gc_percent(pct: i32) -> ItbResult<i32> {
    let s = ffi::syms()?;
    // SAFETY: plain value-in / value-out call.
    Ok(unsafe { (s.ITB_SetGCPercent)(pct) })
}

/// Sets the Go runtime's GOMAXPROCS and returns the previous value.
/// Zero or a negative value queries without changing.
pub fn set_gomaxprocs(n: i32) -> ItbResult<i32> {
    let s = ffi::syms()?;
    // SAFETY: plain value-in / value-out call.
    Ok(unsafe { (s.ITB_SetGOMAXPROCS)(n) })
}

/// Writes the Go runtime's heap profile (pprof format) to `path`
/// after one forced garbage collection. An empty path falls back to
/// the `ITB_MEMPROFILE` environment variable inside libitb3; a path
/// that is still empty, or a file-system failure, fails with
/// [`ItbStatus::BadInput`] and the diagnostic in the error message.
pub fn write_heap_profile(path: impl AsRef<Path>) -> ItbResult<()> {
    let s = ffi::syms()?;
    let text = path
        .as_ref()
        .to_str()
        .ok_or(ItbError::Ffi("path is not valid UTF-8"))?;
    let path_c = CString::new(text).map_err(|_| ItbError::Ffi("path contains NUL"))?;
    // SAFETY: path_c is a live NUL-terminated string for the call.
    check(unsafe { (s.ITB_WriteHeapProfile)(path_c.as_ptr()) })
}

/// Number of `i64` slots [`pool_stats`] returns.
pub fn pool_stats_len() -> ItbResult<usize> {
    let s = ffi::syms()?;
    // SAFETY: value-out call.
    let n = unsafe { (s.ITB_PoolStatsLen)() };
    Ok(if n > 0 { n as usize } else { 0 })
}

/// One snapshot of the library's pool hit / miss counters. Every
/// counter is a monotonically increasing total since library load;
/// difference two snapshots. Slot layout, with `T` the tier count in
/// slot 0: tier `i` holds starter width, checkouts, constructor
/// misses, regrow replacements and bytes allocated at slots
/// `1 + 5*i .. 1 + 5*i + 4`; the scratch byte pool's get / new /
/// regrow / regrow-bytes follow at `1 + 5*T`, and the parallax chunk
/// pool's at `1 + 5*T + 4`.
pub fn pool_stats() -> ItbResult<Vec<i64>> {
    let s = ffi::syms()?;
    let mut v = vec![0i64; pool_stats_len()?];
    let mut written = 0usize;
    // SAFETY: v is writable for v.len() slots; written is a valid
    // out-param.
    check(unsafe { (s.ITB_PoolStats)(v.as_mut_ptr(), v.len(), &mut written) })?;
    v.truncate(written);
    Ok(v)
}

/// Returns the libitb3 library version string.
pub fn version() -> ItbResult<String> {
    let s = ffi::syms()?;
    // SAFETY (both calls): standard probe-then-read over the
    // size-out-param string contract; buffers are live for each call.
    read_cstr(|out, cap, len| unsafe { (s.ITB_Version)(out, cap, len) })
}

/// Two-phase read over the `(out, cap, *out_len)` C-string contract:
/// probe with NULL / 0 for the required capacity, then read and
/// NUL-strip.
fn read_cstr(mut call: impl FnMut(*mut c_char, usize, *mut usize) -> c_int) -> ItbResult<String> {
    let mut need = 0usize;
    let rc = call(std::ptr::null_mut(), 0, &mut need);
    if rc != ItbStatus::Ok as i32 && rc != ItbStatus::BufferTooSmall as i32 {
        return Err(ItbError::from_rc(rc));
    }
    if need <= 1 {
        return Ok(String::new());
    }
    let mut buf = vec![0u8; need];
    check(call(buf.as_mut_ptr().cast(), buf.len(), &mut need))?;
    buf.truncate(need.saturating_sub(1));
    std::str::from_utf8(&buf)
        .map(str::to_owned)
        .map_err(ItbError::Utf8)
}
